File: src/main/java/com/github/xiaolyuh/utils/VirtualFileUtils.java
package com.github.xiaolyuh.utils;

import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.testFramework.LightVirtualFile;

import java.io.IOException;

public class VirtualFileUtils {

    public static VirtualFile createLogVirtualFileFromText(byte[] txtBytes) {
        return createVirtualFileFromText(txtBytes);
    }

    private static VirtualFile createVirtualFileFromText(byte[] txtBytes) {
        try {
            LightVirtualFile lightVirtualFile = new LightVirtualFile("temp-gitFlowPlus.log");
            lightVirtualFile.setBinaryContent(txtBytes);
            return lightVirtualFile;
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

}

File: src/main/kotlin/com/github/xiaolyuh/utils/VirtualFileUtils.kt
package com.github.xiaolyuh.utils

import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.vfs.VfsUtil.findFileByIoFile
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.readText
import java.io.File
import java.nio.file.Files

/**
 * @author yudong
 */
object VirtualFileUtils {

    fun readNewestContent(file: File): String? {
        val virtualFile = findFileByIoFile(file, true)

        if (virtualFile != null && isVirtualFileNewest(virtualFile, file)) {
            return readNewestContent(virtualFile)
        }

        if (!file.exists()) {
            return null
        }

        return Files.readString(file.toPath())
    }

    fun readNewestContent(virtualFile: VirtualFile): String {
        val document = FileDocumentManager.getInstance().getCachedDocument(virtualFile)

        return document?.text ?: virtualFile.readText()
    }

    private fun isVirtualFileNewest(virtualFile: VirtualFile, file: File): Boolean {
        if (!file.exists()) {
            return false
        }

        val timeStamp = virtualFile.timeStamp
        val millis = Files.getLastModifiedTime(file.toPath()).toMillis()
        return timeStamp >= millis
    }
}

